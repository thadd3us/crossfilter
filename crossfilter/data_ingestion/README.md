There are several different CLI programs here than can import data and upsert it into a sqlite database table
with a schema based on //crossfilter/core/schema.py.

# Sample commands

## Ingest GPX file data
```sh
uv run python crossfilter/data_ingestion/gpx/ingest_gpx_files.py ~/Downloads ~/crossfilter/data.sqlite
```

## Ingest Adobe Lightroom data
```sh
# Work laptop:
uv run python crossfilter/data_ingestion/lightroom/ingest_lightroom_catalogs.py \
    ~/personal/non_red ~/crossfilter/data.sqlite \
    --sqlite_db_with_clip_embeddings ~/personal/lightroom_embedding_vectors.sqlite \
    --output_umap_transformation_file ~/crossfilter/clip_umap_transformation.pickle

# Personal laptop
uv run python crossfilter/data_ingestion/lightroom/ingest_lightroom_catalogs.py \
     ~/datasets/lightroom_non_red/ ~/crossfilter/data.sqlite \
    --sqlite_db_with_clip_embeddings ~/monorepo/data/thad/embeddings/images.sqlite \
    --output_umap_transformation_file ~/crossfilter/clip_umap_transformation.pickle

```
