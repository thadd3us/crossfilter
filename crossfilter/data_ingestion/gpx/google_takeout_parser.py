"""Google Takeout location history parsing functions that convert JSON files to DataFrames."""

import hashlib
import json
import logging
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

import pandas as pd

from crossfilter.core.bucketing import add_geo_h3_bucket_columns
from crossfilter.core.schema import DataType
from crossfilter.core.schema import SchemaColumns as C

logger = logging.getLogger(__name__)


def generate_uuid_from_components(
    first_point_timestamp: datetime,
    first_point_lat: float,
    first_point_lon: float,
    point_timestamp: datetime,
    point_lat: float,
    point_lon: float,
    point_type: DataType,
) -> bytes:
    """Generate a deterministic UUID from the specified components."""
    # Create a string representation of all components
    components = [
        first_point_timestamp.isoformat(),
        str(first_point_lat),
        str(first_point_lon),
        point_timestamp.isoformat(),
        str(point_lat),
        str(point_lon),
        point_type.value,
    ]

    # Create a deterministic hash from the components
    component_string = "|".join(components)
    hash_digest = hashlib.sha256(component_string.encode("utf-8")).digest()
    return hash_digest[:16]


def generate_uuid_for_row(
    row: pd.Series,
    first_point_timestamp: datetime,
    first_point_lat: float,
    first_point_lon: float,
) -> bytes:
    """Generate a UUID for a single DataFrame row."""
    return generate_uuid_from_components(
        first_point_timestamp,
        first_point_lat,
        first_point_lon,
        row[C.TIMESTAMP_UTC].to_pydatetime(),
        row[C.GPS_LATITUDE],
        row[C.GPS_LONGITUDE],
        DataType.GOOGLE_TAKEOUT_LOCATION,
    )


def parse_takeout_timestamp(timestamp_str: str) -> datetime:
    """
    Parse Google Takeout timestamp string to datetime object.
    
    Google Takeout uses ISO 8601 format timestamps like '2022-01-12T17:18:24.190Z'
    """
    try:
        # Parse ISO 8601 timestamp
        dt = datetime.fromisoformat(timestamp_str.replace('Z', '+00:00'))
        # Ensure UTC timezone
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        return dt
    except ValueError as e:
        raise ValueError(f"Failed to parse timestamp '{timestamp_str}'") from e


def convert_e7_coordinate(e7_coord: int) -> float:
    """
    Convert Google's E7 coordinate format to decimal degrees.
    
    Google stores coordinates multiplied by 10^7 as integers.
    For example, latitudeE7: 414216106 becomes 41.4216106 degrees.
    """
    return e7_coord / 1e7


def parse_location_record(record: dict[str, Any], source_filename: str) -> dict[str, Any] | None:
    """
    Parse a single location record from Google Takeout Records.json.
    
    Returns None if the record is invalid or missing required fields.
    """
    # Check for required fields
    if "timestamp" not in record:
        logger.debug("Skipping record without timestamp")
        return None
        
    if "latitudeE7" not in record or "longitudeE7" not in record:
        logger.debug("Skipping record without coordinates")
        return None

    try:
        timestamp = parse_takeout_timestamp(record["timestamp"])
        latitude = convert_e7_coordinate(record["latitudeE7"])
        longitude = convert_e7_coordinate(record["longitudeE7"])
        
        # Validate coordinate ranges
        if not (-90 <= latitude <= 90):
            logger.warning(f"Invalid latitude {latitude}, skipping record")
            return None
        if not (-180 <= longitude <= 180):
            logger.warning(f"Invalid longitude {longitude}, skipping record")
            return None
        
        return {
            C.DATA_TYPE: DataType.GOOGLE_TAKEOUT_LOCATION,
            C.SOURCE_FILE: source_filename,
            C.TIMESTAMP_MAYBE_TIMEZONE_AWARE: record["timestamp"],
            C.TIMESTAMP_UTC: timestamp,
            C.GPS_LATITUDE: latitude,
            C.GPS_LONGITUDE: longitude,
        }
        
    except (ValueError, KeyError) as e:
        logger.debug(f"Failed to parse location record: {e}")
        return None


def load_google_takeout_records_to_df(records_json_path: Path) -> pd.DataFrame:
    """
    Parse a Google Takeout Records.json file and convert it to a DataFrame conforming to the schema.

    Args:
        records_json_path: Path to the Records.json file from Google Takeout

    Returns:
        DataFrame with location data as GPX trackpoints

    Raises:
        FileNotFoundError: If the JSON file doesn't exist
        ValueError: If the JSON file is invalid or has unexpected structure
    """
    if not records_json_path.exists():
        raise FileNotFoundError(f"Google Takeout Records.json file not found: {records_json_path}")

    try:
        with open(records_json_path, encoding="utf-8") as f:
            data = json.load(f)
    except json.JSONDecodeError as e:
        raise ValueError(f"Failed to parse JSON file {records_json_path}") from e
    except Exception as e:
        raise ValueError(f"Failed to read JSON file {records_json_path}") from e

    if not isinstance(data, dict) or "locations" not in data:
        raise ValueError(f"Invalid Google Takeout Records.json structure: missing 'locations' array in {records_json_path}")

    locations = data["locations"]
    if not isinstance(locations, list):
        raise ValueError(f"Invalid Google Takeout Records.json structure: 'locations' is not an array in {records_json_path}")

    # Collect all valid location records
    records: list[dict] = []
    source_filename = records_json_path.name

    logger.info(f"Processing {len(locations)} location records from {records_json_path}")
    
    for i, location in enumerate(locations):
        if not isinstance(location, dict):
            logger.debug(f"Skipping non-dict location record at index {i}")
            continue
            
        parsed_record = parse_location_record(location, source_filename)
        if parsed_record:
            records.append(parsed_record)

    if not records:
        logger.warning(f"No valid location records found in {records_json_path}")
        # Return empty DataFrame with correct schema
        df = pd.DataFrame(
            columns=[
                C.UUID,
                C.DATA_TYPE,
                C.NAME,
                C.SOURCE_FILE,
                C.TIMESTAMP_MAYBE_TIMEZONE_AWARE,
                C.TIMESTAMP_UTC,
                C.GPS_LATITUDE,
                C.GPS_LONGITUDE,
            ]
        )
    else:
        # Create DataFrame
        df = pd.DataFrame(records)

    # Convert columns to proper types
    if not df.empty:
        df[C.TIMESTAMP_UTC] = pd.to_datetime(df[C.TIMESTAMP_UTC], utc=True)
        df[C.GPS_LATITUDE] = df[C.GPS_LATITUDE].astype(float)
        df[C.GPS_LONGITUDE] = df[C.GPS_LONGITUDE].astype(float)
        df = df.sort_values(
            by=[C.TIMESTAMP_UTC, C.DATA_TYPE, C.GPS_LATITUDE, C.GPS_LONGITUDE]
        )

        # Find the first point for UUID generation (first chronologically)
        first_row = df.iloc[0]
        first_point_timestamp = first_row[C.TIMESTAMP_UTC].to_pydatetime()
        first_point_lat = first_row[C.GPS_LATITUDE]
        first_point_lon = first_row[C.GPS_LONGITUDE]

        # Generate UUIDs for all rows using vectorized operation
        df[C.UUID] = df.apply(
            lambda row: generate_uuid_for_row(
                row, first_point_timestamp, first_point_lat, first_point_lon
            ),
            axis=1,
        )

    # Set index name
    df.index.name = C.DF_ID

    # Add H3 spatial index columns (parallelized computation)
    if not df.empty and C.GPS_LATITUDE in df.columns and C.GPS_LONGITUDE in df.columns:
        add_geo_h3_bucket_columns(df)
        logger.debug(f"Added H3 columns to {len(df)} points from {records_json_path}")

    logger.info(f"Loaded {len(df)} location points from {records_json_path}")

    return df