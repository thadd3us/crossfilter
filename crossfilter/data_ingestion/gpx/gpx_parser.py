"""GPX file parsing functions that convert GPX files to DataFrames."""

import hashlib
import logging
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import List, Optional

import gpxpy
import pandas as pd
from gpxpy.gpx import GPXTrackPoint, GPXWaypoint

from crossfilter.core.bucketing import add_geo_h3_bucket_columns
from crossfilter.core.schema import DataType, SchemaColumns as C

logger = logging.getLogger(__name__)


def generate_uuid_from_components(
    first_trackpoint_timestamp: datetime,
    first_trackpoint_lat: float,
    first_trackpoint_lon: float,
    point_timestamp: datetime,
    point_lat: float,
    point_lon: float,
    point_type: DataType,
) -> str:
    """Generate a deterministic UUID from the specified components."""
    # Create a string representation of all components
    components = [
        first_trackpoint_timestamp.isoformat(),
        str(first_trackpoint_lat),
        str(first_trackpoint_lon),
        point_timestamp.isoformat(),
        str(point_lat),
        str(point_lon),
        point_type.value,
    ]

    # Create a deterministic hash from the components
    component_string = "|".join(components)
    hash_digest = hashlib.sha256(component_string.encode("utf-8")).digest()

    # Create a UUID from the hash
    return str(uuid.UUID(bytes=hash_digest[:16]))


def generate_uuid_for_row(
    row: pd.Series,
    first_point_timestamp: datetime,
    first_point_lat: float,
    first_point_lon: float,
) -> str:
    """Generate a UUID for a single DataFrame row."""
    return generate_uuid_from_components(
        first_point_timestamp,
        first_point_lat,
        first_point_lon,
        row[C.TIMESTAMP_UTC].to_pydatetime(),
        row[C.GPS_LATITUDE],
        row[C.GPS_LONGITUDE],
        DataType(row[C.DATA_TYPE]),
    )


def load_gpx_file_to_df(gpx_file_path: Path) -> pd.DataFrame:
    """
    Parse a GPX file and convert it to a DataFrame conforming to the schema.

    Args:
        gpx_file_path: Path to the GPX file

    Returns:
        DataFrame with GPX trackpoints and waypoints

    Raises:
        FileNotFoundError: If the GPX file doesn't exist
        ValueError: If the GPX file is invalid or empty
    """
    if not gpx_file_path.exists():
        raise FileNotFoundError(f"GPX file not found: {gpx_file_path}")

    try:
        with open(gpx_file_path, "r", encoding="utf-8") as f:
            gpx = gpxpy.parse(f)
    except Exception as e:
        raise ValueError(f"Failed to parse GPX file {gpx_file_path}: {e}")

    # Get file size
    file_size = gpx_file_path.stat().st_size

    # Collect all points first (without UUIDs)
    records: List[dict] = []

    # Process all trackpoints
    for track in gpx.tracks:
        for segment in track.segments:
            for point in segment.points:
                if point.time and point.latitude and point.longitude:
                    records.append(
                        {
                            C.DATA_TYPE: DataType.GPX_TRACKPOINT,
                            C.SOURCE_FILE: str(gpx_file_path.name),
                            C.TIMESTAMP_MAYBE_TIMEZONE_AWARE: point.time.isoformat(),
                            C.TIMESTAMP_UTC: (
                                point.time
                                if point.time.tzinfo is not None
                                else point.time.replace(tzinfo=timezone.utc)
                            ),
                            C.GPS_LATITUDE: point.latitude,
                            C.GPS_LONGITUDE: point.longitude,
                        }
                    )

    # Process all waypoints
    for waypoint in gpx.waypoints:
        if waypoint.time and waypoint.latitude and waypoint.longitude:
            records.append(
                {
                    C.DATA_TYPE: DataType.GPX_WAYPOINT,
                    C.NAME: waypoint.name,
                    C.CAPTION: waypoint.comment or waypoint.description,
                    C.SOURCE_FILE: str(gpx_file_path.name),
                    C.TIMESTAMP_MAYBE_TIMEZONE_AWARE: waypoint.time.isoformat(),
                    C.TIMESTAMP_UTC: (
                        waypoint.time
                        if waypoint.time.tzinfo is not None
                        else waypoint.time.replace(tzinfo=timezone.utc)
                    ),
                    C.GPS_LATITUDE: waypoint.latitude,
                    C.GPS_LONGITUDE: waypoint.longitude,
                }
            )

    if not records:
        logger.warning(f"No valid trackpoints or waypoints found in {gpx_file_path}")
        # Return empty DataFrame with correct schema
        df = pd.DataFrame(
            columns=[
                C.UUID_STRING,
                C.DATA_TYPE,
                C.NAME,
                C.SOURCE_FILE,
                C.TIMESTAMP_MAYBE_TIMEZONE_AWARE,
                C.TIMESTAMP_UTC,
                C.GPS_LATITUDE,
                C.GPS_LONGITUDE,
            ]
        )
    else:
        # Create DataFrame
        df = pd.DataFrame(records)

    # Convert TIMESTAMP_UTC to proper datetime with UTC timezone
    df[C.TIMESTAMP_UTC] = pd.to_datetime(df[C.TIMESTAMP_UTC], utc=True)
    df[C.GPS_LATITUDE] = df[C.GPS_LATITUDE].astype(float)
    df[C.GPS_LONGITUDE] = df[C.GPS_LONGITUDE].astype(float)
    df = df.sort_values(
        by=[C.TIMESTAMP_UTC, C.DATA_TYPE, C.GPS_LATITUDE, C.GPS_LONGITUDE]
    )

    if not df.empty:
        # Find the first point for UUID generation (first chronologically)
        first_row = df.iloc[0]
        first_point_timestamp = first_row[C.TIMESTAMP_UTC].to_pydatetime()
        first_point_lat = first_row[C.GPS_LATITUDE]
        first_point_lon = first_row[C.GPS_LONGITUDE]

        # Generate UUIDs for all rows using vectorized operation
        df[C.UUID_STRING] = df.apply(
            lambda row: generate_uuid_for_row(
                row, first_point_timestamp, first_point_lat, first_point_lon
            ),
            axis=1,
        )

    # Set index name
    df.index.name = C.DF_ID

    # Add H3 spatial index columns for each individual GPX file (parallelized computation)
    if not df.empty and C.GPS_LATITUDE in df.columns and C.GPS_LONGITUDE in df.columns:
        add_geo_h3_bucket_columns(df)
        logger.debug(f"Added H3 columns to {len(df)} points from {gpx_file_path}")

    logger.info(f"Loaded {len(df)} points from {gpx_file_path}")

    return df
