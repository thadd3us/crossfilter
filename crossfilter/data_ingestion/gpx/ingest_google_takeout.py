"""CLI program for ingesting Google Takeout Records.json files into parquet."""

import logging
from pathlib import Path

import pandas as pd
import typer
from tqdm.contrib.concurrent import process_map

from crossfilter.core.schema import SchemaColumns as C
from crossfilter.data_ingestion.gpx.google_takeout_parser import load_google_takeout_records_to_df

logger = logging.getLogger(__name__)
typer.main.get_command_name = lambda name: name
app = typer.Typer(pretty_exceptions_show_locals=False)


def find_takeout_files(base_dir: Path) -> list[Path]:
    """Find all Google Takeout Records.json files recursively in the base directory."""
    if not base_dir.exists():
        raise FileNotFoundError(f"Base directory not found: {base_dir}")

    if not base_dir.is_dir():
        raise ValueError(f"Base directory is not a directory: {base_dir}")

    # Look for Records.json files (Google Takeout location history)
    takeout_files = list(base_dir.rglob("Records.json"))
    logger.info(f"Found {len(takeout_files)} Google Takeout Records.json files in {base_dir}")

    return takeout_files


def process_single_takeout_file(takeout_file: Path) -> pd.DataFrame:
    """Process a single Google Takeout Records.json file and return a DataFrame."""
    try:
        return load_google_takeout_records_to_df(takeout_file)
    except Exception as e:
        logger.error(f"Failed to process {takeout_file}: {e}")
        # Return empty DataFrame with correct schema on error
        return pd.DataFrame(
            columns=[
                C.UUID,
                C.DATA_TYPE,
                C.SOURCE_FILE,
                C.TIMESTAMP_MAYBE_TIMEZONE_AWARE,
                C.TIMESTAMP_UTC,
                C.GPS_LATITUDE,
                C.GPS_LONGITUDE,
            ]
        )


@app.command()
def main(
    base_dir: Path = typer.Argument(
        ..., help="Directory to recursively search for Google Takeout Records.json files"
    ),
    output_parquet: Path = typer.Argument(..., help="Parquet file to write to"),
    max_workers: int = typer.Option(None, help="Maximum number of parallel workers"),
) -> None:
    """Ingest Google Takeout Records.json files into parquet format."""
    logging.basicConfig(level=logging.INFO)

    # Find all Takeout files
    takeout_files = find_takeout_files(base_dir)

    if not takeout_files:
        logger.info("No Google Takeout Records.json files found")
        return

    # Process files in parallel
    logger.info(f"Processing {len(takeout_files)} Google Takeout files...")
    dataframes = process_map(
        process_single_takeout_file,
        takeout_files,
        max_workers=max_workers,
        desc="Processing Takeout files",
        chunksize=1,
    )

    # Filter out empty DataFrames
    non_empty_dataframes = [df for df in dataframes if not df.empty]

    if not non_empty_dataframes:
        logger.info("No valid data found in Google Takeout files")
        return

    # Concatenate all DataFrames
    logger.info(f"Concatenating {len(non_empty_dataframes)} DataFrames...")
    combined_df = pd.concat(non_empty_dataframes, ignore_index=True)

    combined_df = combined_df.drop_duplicates()
    combined_df = combined_df.sort_values([C.TIMESTAMP_UTC, C.UUID], ignore_index=True)

    if combined_df[C.UUID].duplicated().any():
        logger.warning(
            "Duplicate UUIDs found in data. This should not happen if the data is clean."
        )
        dupes = combined_df[C.UUID].duplicated(keep=False)
        logger.info(f"Dupes: {combined_df[dupes][C.SOURCE_FILE].value_counts()}")
        raise ValueError("Duplicate UUIDs found in data.")

    logger.info(
        f"Total records: {len(combined_df)} (H3 columns already computed per-file in parallel)"
    )
    combined_df[C.UUID] = combined_df[C.UUID].astype('object')
    combined_df = combined_df.set_index(C.UUID, verify_integrity=True)

    logger.info(f"Writing to parquet file: {output_parquet}")
    combined_df.to_parquet(
        output_parquet,
        engine="pyarrow",
        index=True,
    )

    logger.info("Google Takeout ingestion completed successfully")


if __name__ == "__main__":
    app()