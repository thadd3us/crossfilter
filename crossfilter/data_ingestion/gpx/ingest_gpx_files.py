"""CLI program for ingesting GPX files into SQLite database."""

import logging
from pathlib import Path
from typing import List

import pandas as pd
import typer
from tqdm.contrib.concurrent import process_map
from crossfilter.core.schema import SchemaColumns as C

from crossfilter.core.schema import SchemaColumns
from crossfilter.data_ingestion.gpx.gpx_parser import load_gpx_file_to_df
from crossfilter.data_ingestion.sqlite_utils import upsert_dataframe_to_sqlite

logger = logging.getLogger(__name__)


def find_gpx_files(base_dir: Path) -> List[Path]:
    """Find all GPX files recursively in the base directory."""
    if not base_dir.exists():
        raise FileNotFoundError(f"Base directory not found: {base_dir}")

    if not base_dir.is_dir():
        raise ValueError(f"Base directory is not a directory: {base_dir}")

    gpx_files = list(base_dir.rglob("*.gpx"))
    logger.info(f"Found {len(gpx_files)} GPX files in {base_dir}")

    return gpx_files


def process_single_gpx_file(gpx_file: Path) -> pd.DataFrame:
    """Process a single GPX file and return a DataFrame."""
    try:
        return load_gpx_file_to_df(gpx_file)
    except Exception as e:
        logger.error(f"Failed to process {gpx_file}: {e}")
        # Return empty DataFrame with correct schema on error
        return pd.DataFrame(
            columns=[
                SchemaColumns.UUID_STRING,
                SchemaColumns.DATA_TYPE,
                SchemaColumns.NAME,
                SchemaColumns.CAPTION,
                SchemaColumns.SOURCE_FILE,
                SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE,
                SchemaColumns.TIMESTAMP_UTC,
                SchemaColumns.GPS_LATITUDE,
                SchemaColumns.GPS_LONGITUDE,
                SchemaColumns.RATING_0_TO_5,
                SchemaColumns.SIZE_IN_BYTES,
            ]
        )


def main(
    base_dir: Path = typer.Argument(
        ..., help="Directory to recursively search for GPX files"
    ),
    destination_sqlite_db: Path = typer.Argument(
        ..., help="SQLite database file to upsert into"
    ),
    destination_table: str = typer.Option("data", help="Database table name"),
    max_workers: int = typer.Option(None, help="Maximum number of parallel workers"),
) -> None:
    """Ingest GPX files into SQLite database."""
    logging.basicConfig(level=logging.INFO)

    # Find all GPX files
    gpx_files = find_gpx_files(base_dir)

    if not gpx_files:
        logger.info("No GPX files found")
        return

    # Process files in parallel
    logger.info(f"Processing {len(gpx_files)} GPX files...")
    dataframes = process_map(
        process_single_gpx_file,
        gpx_files,
        max_workers=max_workers,
        desc="Processing GPX files",
        chunksize=1,
    )

    # Filter out empty DataFrames
    non_empty_dataframes = [df for df in dataframes if not df.empty]

    if not non_empty_dataframes:
        logger.info("No valid data found in GPX files")
        return

    # Concatenate all DataFrames
    logger.info(f"Concatenating {len(non_empty_dataframes)} DataFrames...")
    combined_df = pd.concat(non_empty_dataframes, ignore_index=True)

    combined_df = combined_df.drop_duplicates()

    if combined_df[C.UUID_STRING].duplicated().any():
        logger.warning("Duplicate UUIDs found in data")
        # assert False
        combined_df = combined_df.drop_duplicates(subset=[C.UUID_STRING])

    logger.info(
        f"Total records: {len(combined_df)} (H3 columns computed per-file in parallel)"
    )

    # Upsert to database
    upsert_dataframe_to_sqlite(combined_df, destination_sqlite_db, destination_table)

    logger.info("GPX ingestion completed successfully")


if __name__ == "__main__":
    typer.run(main)
