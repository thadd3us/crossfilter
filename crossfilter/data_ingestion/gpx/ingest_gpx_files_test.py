"""Tests for GPX files ingestion CLI program."""

from pathlib import Path

import pandas as pd
import pytest
from sqlalchemy import create_engine
from syrupy.assertion import SnapshotAssertion

from crossfilter.core.schema import DataType, SchemaColumns
from crossfilter.data_ingestion.gpx.ingest_gpx_files import (
    find_gpx_files,
    main,
    process_single_gpx_file,
)
from crossfilter.data_ingestion.sqlite_utils import query_sqlite_to_dataframe

gpx_content = """<?xml version="1.0" encoding="UTF-8"?>
<gpx version="1.1" creator="test">
    <trk>
        <name>Test Track</name>
        <trkseg>
            <trkpt lat="37.7749" lon="-122.4194">
                <time>2023-01-01T12:00:00Z</time>
            </trkpt>
        </trkseg>
    </trk>
</gpx>"""


@pytest.fixture
def gpx_file(tmp_path: Path) -> Path:
    """Create a test GPX file with given content."""
    gpx_file = tmp_path / "test.gpx"
    gpx_file.write_text(gpx_content)
    return gpx_file


def test_find_gpx_files_nonexistent_directory() -> None:
    """Test finding GPX files in non-existent directory."""
    nonexistent_dir = Path("/nonexistent/directory")

    with pytest.raises(FileNotFoundError):
        find_gpx_files(nonexistent_dir)


def test_find_gpx_files_not_directory(tmp_path: Path) -> None:
    """Test finding GPX files when path is not a directory."""
    test_file = tmp_path / "test.txt"
    test_file.write_text("test")

    with pytest.raises(ValueError):
        find_gpx_files(test_file)


def test_find_gpx_files_empty_directory(tmp_path: Path) -> None:
    """Test finding GPX files in empty directory."""
    gpx_files = find_gpx_files(tmp_path)
    assert gpx_files == []


def test_find_gpx_files_with_files(tmp_path: Path) -> None:
    """Test finding GPX files in directory with files."""
    # Create some GPX files
    (tmp_path / "file1.gpx").write_text("content1")
    (tmp_path / "file2.gpx").write_text("content2")
    (tmp_path / "other.txt").write_text("content3")

    # Create subdirectory with GPX file
    subdir = tmp_path / "subdir"
    subdir.mkdir()
    (subdir / "file3.gpx").write_text("content4")

    gpx_files = find_gpx_files(tmp_path)

    assert len(gpx_files) == 3
    assert tmp_path / "file1.gpx" in gpx_files
    assert tmp_path / "file2.gpx" in gpx_files
    assert subdir / "file3.gpx" in gpx_files


def test_process_single_gpx_file_valid(tmp_path: Path, gpx_file: Path) -> None:
    """Test processing a valid GPX file."""
    df = process_single_gpx_file(gpx_file)

    assert len(df) == 1
    assert df.iloc[0][SchemaColumns.DATA_TYPE] == DataType.GPX_TRACKPOINT
    assert df.iloc[0][SchemaColumns.GPS_LATITUDE] == 37.7749
    assert df.iloc[0][SchemaColumns.GPS_LONGITUDE] == -122.4194


def test_process_single_gpx_file_invalid(tmp_path: Path) -> None:
    """Test processing an invalid GPX file."""
    invalid_file = tmp_path / "invalid.gpx"
    invalid_file.write_text("This is not a valid GPX file")

    df = process_single_gpx_file(invalid_file)

    # Should return empty DataFrame with correct schema
    assert len(df) == 0
    assert SchemaColumns.DATA_TYPE in df.columns


def test_thad_ingest_dev_data(
    gpx_file: Path, tmp_path: Path, snapshot: SnapshotAssertion
) -> None:
    main(
        base_dir=gpx_file.parent,
        max_workers=1,
        output_parquet=tmp_path / "data.parquet",
        output_sqlite_db=None,
        output_table=None,
    )

    df = pd.read_parquet(tmp_path / "data.parquet")
    # Reset index to include UUID column in the records
    df = df.reset_index()
    assert df.to_dict(orient="records") == snapshot

    # engine = create_engine(f"sqlite:///{tmp_path / 'data.sqlite'}")
    # df = query_sqlite_to_dataframe(engine, "SELECT * FROM data")
    # assert df.to_dict(orient="records") == snapshot
