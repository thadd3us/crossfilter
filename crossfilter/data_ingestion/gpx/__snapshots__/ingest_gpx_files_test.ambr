# serializer version: 1
# name: test_thad_ingest_dev_data
  list([
    dict({
      'BUCKETED_H3_L0': '8029fffffffffff',
      'BUCKETED_H3_L1': '81283ffffffffff',
      'BUCKETED_H3_L10': '8a283082800ffff',
      'BUCKETED_H3_L11': '8b283082800bfff',
      'BUCKETED_H3_L12': '8c283082800b3ff',
      'BUCKETED_H3_L13': '8d283082800b3bf',
      'BUCKETED_H3_L14': '8e283082800b397',
      'BUCKETED_H3_L15': '8f283082800b390',
      'BUCKETED_H3_L2': '822837fffffffff',
      'BUCKETED_H3_L3': '832830fffffffff',
      'BUCKETED_H3_L4': '8428309ffffffff',
      'BUCKETED_H3_L5': '85283083fffffff',
      'BUCKETED_H3_L6': '86283082fffffff',
      'BUCKETED_H3_L7': '872830828ffffff',
      'BUCKETED_H3_L8': '8828308281fffff',
      'BUCKETED_H3_L9': '89283082803ffff',
      'DATA_TYPE': 'GPX_TRACKPOINT',
      'GPS_LATITUDE': 37.7749,
      'GPS_LONGITUDE': -122.4194,
      'SOURCE_FILE': 'test.gpx',
      'TIMESTAMP_MAYBE_TIMEZONE_AWARE': '2023-01-01T12:00:00+00:00',
      'TIMESTAMP_UTC': Timestamp('2023-01-01 12:00:00+0000', tz='UTC'),
      'UUID': b'\xf5\x0c\xa1\xb8N\xf2Rw\xf2\x04\xea\xd8\xff\x90\x01\x0b',
    }),
  ])
# ---
